os_stub/cryptlib_openssl/pk/ec.c

```diff
@@ -846,3 +846,221 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
+    *kinvp = k;
+    ret = 1;
+
+err:
+    if (!ret) {
+        BN_clear_free(k);
+        BN_clear_free(r);
+    }
+
+    BN_CTX_free(ctx);
+    EC_POINT_free(tmp_point);
+    BN_clear_free(X);
+    return ret;
+}
+
+/**
+ * Carries out the EC-DSA signature with caller input random function. This API can be used for FIPS test.
+ *
+ * @param[in]       ec_context    Pointer to EC context for signature generation.
+ * @param[in]       hash_nid      hash NID
+ * @param[in]       message_hash  Pointer to octet message hash to be signed.
+ * @param[in]       hash_size     Size of the message hash in bytes.
+ * @param[out]      signature     Pointer to buffer to receive EC-DSA signature.
+ * @param[in, out]  sig_size      On input, the size of signature buffer in bytes.
+ *                                On output, the size of data returned in signature buffer in bytes.
+ * @param[in]       random_func   random number function
+ *
+ * @retval  true   signature successfully generated in EC-DSA.
+ * @retval  false  signature generation failed.
+ * @retval  false  sig_size is too small.
+ **/
+bool libspdm_ecdsa_sign_ex(void *ec_context, size_t hash_nid,
+                           const uint8_t *message_hash, size_t hash_size,
+                           uint8_t *signature, size_t *sig_size,
+                           int (*random_func)(void *, unsigned char *, size_t))
+{
+    EC_KEY *ec_key;
+    ECDSA_SIG *ecdsa_sig;
+    int32_t openssl_nid;
+    uint8_t half_size;
+    BIGNUM *bn_r;
+    BIGNUM *bn_s;
+    int r_size;
+    int s_size;
+    /* hardcode random number*/
+    uint8_t random[32];
+
+    BIGNUM *kinv = NULL;
+    BIGNUM *rp = NULL;
+
+    kinv = NULL;
+    rp = NULL;
+
+    if (ec_context == NULL || message_hash == NULL) {
+        return false;
+    }
+
+    if (signature == NULL) {
+        return false;
+    }
+
+    ec_key = (EC_KEY *)ec_context;
+    openssl_nid = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec_key));
+    switch (openssl_nid) {
+    case NID_X9_62_prime256v1:
+        half_size = 32;
+        break;
+    case NID_secp384r1:
+        half_size = 48;
+        break;
+    case NID_secp521r1:
+        half_size = 66;
+        break;
+    default:
+        return false;
+    }
+    if (*sig_size < (size_t)(half_size * 2)) {
+        *sig_size = half_size * 2;
+        return false;
+    }
+    *sig_size = half_size * 2;
+    libspdm_zero_mem(signature, *sig_size);
+
+    switch (hash_nid) {
+    case LIBSPDM_CRYPTO_NID_SHA256:
+        if (hash_size != LIBSPDM_SHA256_DIGEST_SIZE) {
+            return false;
+        }
+        break;
+
+    case LIBSPDM_CRYPTO_NID_SHA384:
+        if (hash_size != LIBSPDM_SHA384_DIGEST_SIZE) {
+            return false;
+        }
+        break;
+
+    case LIBSPDM_CRYPTO_NID_SHA512:
+        if (hash_size != LIBSPDM_SHA512_DIGEST_SIZE) {
+            return false;
+        }
+        break;
+
+    default:
+        return false;
+    }
+
+    /*hard code random number for ecdsa sign*/
```

Comment:
`hard code random number for ecdsa sign` -> `retrieve random number for ecdsa sign`